lib/rspec/rails/matchers/send_email.rb

```diff
@@ -0,0 +1,125 @@
+# frozen_string_literal: true
+
+module RSpec
+  module Rails
+    module Matchers
+      # @api private
+      #
+      # Matcher class for `send_email`. Should not be instantiated directly.
+      #
+      # @see RSpec::Rails::Matchers#send_email
+      class SendEmail < RSpec::Rails::Matchers::BaseMatcher
+        INSPECT_EMAIL_ATTRIBUTES = %i[subject from to cc bcc].freeze
+
+        def initialize(criteria)
+          @criteria = criteria
+        end
+
+        # @private
+        def supports_value_expectations?
+          false
+        end
+
+        # @private
+        def supports_block_expectations?
+          true
+        end
+
+        def matches?(block)
+          define_matched_emails(block)
+
+          @matched_emails.one?
+        end
+
+        private
+
+        def diffable?
+          true
+        end
+
+        def match_when_negated(block)
```

Comment:
Would it be possible to do this with `does_not_match?`?
```suggestion
        def does_not_match?(block)
```

I think of uses with compound matching like:
```ruby
expect {}
 .to send_email(to: 'foo@...')
 .and send_email(to: 'bar@...')
```
and if in certain cases when the first or the second matcher fails there would be a difference in the output.

Reply:
Good catch! Actually, the compound version didn't work at all, because the `failure_message` defined below should be public.

I didn't see the difference between the usage of `match_when_negated` and `does_not_match?`. At least my tests did not expose any difference. Moreover, I realized the default implementation of `does_not_match?` is good enough. So removing it.

Another fun thing is that `failure_message_when_negated` can be private while `failure_message` should be public. But I move them both to public space for consistency as the other matchers do.